### PR TITLE
Update Python to 3.11.9

### PR DESCRIPTION
**Pull request by daniandtheweb**
This PR updates Python to version 3.11.9.

### CODE REVIEW

**Comment by daniandtheweb** [comment]
I've added a patch that allows to build pytorch-vision on Python 3.11, thanks @jeroen-mostert for the suggestion here: https://github.com/lamikr/rocm_sdk_builder/issues/66#issuecomment-2169018614.

This PR should be ready to merge as I don't find any issue compared to 3.9.
If someone else could test it too it would be great.

File: patches/rocm-6.1.2/pytorch_vision/0003-pytorch-vision-build-fix-for-python311.patch [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+diff --git a/setup.py b/setup.py
+index ce1cd90ca0..1df620fe83 100644
+--- a/setup.py
++++ b/setup.py
+@@ -383,9 +383,7 @@ def get_extensions():
+         ffmpeg_include_dir = os.path.join(ffmpeg_root, "include")
+         ffmpeg_library_dir = os.path.join(ffmpeg_root, "lib")
+ 
+-        gcc = os.environ.get("CC", shutil.which("gcc"))
```
Comment (jeroen-mostert) [inline patches/rocm-6.1.2/pytorch_vision/0003-pytorch-vision-build-fix-for-python311.patch:22]: This is a little crude; my suggestion was actually (apologies for not making this a pull) using the `is_rocm_pytorch` flag set earlier:

```python
platform_tag = ""
if not is_rocm_pytorch:
    platform_tag = subprocess.run(...
    ...

if platform_tag:
    ...
```

Of course stubbing out the whole `platform_tag` logic also works if we know we never want to build this with gcc, but then all of the logic should simply be deleted, including the `if` block (`platform_tag` is not used outside this block, this appears to be logic specifically for Debian only). This should definitely be tested on Debian-derived systems to ensure it doesn't actually break anything that Debian needs. If the build outputs `Building torchvision without ffmpeg support` even when ffmpeg is installed that would be a bad sign.

Using the compiler to get the architecture seems suspect to begin with and there's probably a more elegant way that doesn't involve invoking gcc.

Comment (daniandtheweb) [reply patches/rocm-6.1.2/pytorch_vision/0003-pytorch-vision-build-fix-for-python311.patch:22]: Ohh, I didn't understand how you were thinking to implement it so I just went straight to the fastest fix.
I'll test it later, thanks for the explanation.

**Comment by lamikr** [comment]
Sorry, I have been trying to trigger the pytorch audio error for last couple of days with the rocm-6.1.2 and python 3.9 so have not had time to test this yet. I will put one test build running for this later on today.

**Comment by lamikr** [comment]
Little related question. What version of numpy you have on your python 3.11 version of rocm installed?

**Comment by daniandtheweb** [comment]
> Little related question. What version of numpy you have on your python 3.11 version of rocm installed?

Version 2.0.0

**Comment by lamikr** [comment]
I started testing now the update to 3.11.9 by doing my first build. I merged the patches now manually.
@daniandtheweb would it be possible for you to rebase/compine all changes to 2 patches and force push them to this pull request branch. (So that there would not be merge commits)

@daniandtheweb @jeroen-mostert Have you tested this with stable diffusion next? As python major version and it's related dependencies update can affect potentially so many different places, I want to spend some time testing this and python 3.9 build against each others.

**Comment by daniandtheweb** [comment]
Sure, I'll rebase it right now.

I've been testing this build for some days now with both ComfyUI and SDnext and I haven't found any specific issue, every issue I had was already present on the 3.9 build and it's most likely related to the fact that RDNA1 isn't officially supported.

**Comment by daniandtheweb** [comment]
@lamikr let me know if this seems right.

**Comment by lamikr** [comment]
I went through the changes and they look good. Could you however if you could squash these 6 patches to 2 or 3 and then use git force-push to recreate this pull-request? This will make the git history much more clear to see what kind of changes have been needed for python update.  I tested that these commands is one way to squash them in your repo for example to 2 commits.

```
git checout python_updates
git rebase -i d6b6ebe466cb1cc305bb10ca928be1adf5cff4e6

```

It opens the editor showing the patches, I edited it to look like this: (I changed 4 from 5 pick words that it originally shows to s) and then accepted the change in git editor.

```
pick f793497 Fix build on Python 3.11 patch
s 836cfcc Fix install on Python 3.11
s 240e27b Improve torchvision patch
s 336de2c Support Python 3.11 for AMDMIGraphX
s 26a1d4a Fix patch author
```
After this it opened the another editor window where I could write the new commit message in style "rocm sdk app build fixes for python 3.11". Once saved, your git log history should show only 2 commits. And then you can just force repush to create a new request for same branch. (Or push to new branch to create second pr)

`git push -f origin HEAD:python_update`

**Comment by daniandtheweb** [comment]
Done, and thanks for the guide, it's my first time actually using the git rebase command and that helped me a lot.

**Comment by lamikr** [comment]
Thanks for fast reply, everything looked good and I rebased my test build to that and restarted the test.
I also included the ffmpeg7 pytorch audio patch to same build.

**Comment by lamikr** [comment]
Unfortunately a new problem on Fedora 40.

```

-- OS: fedora 40
-- Tensile script: /home/lamikr/own/fedora/rocm/src/sdk/rocm_sdk_builder_612/builddir/023_02_rocBLAS/virtualenv/lib/python3.11/site-packages/Tensile/bin/TensileCreateLibrary
-- Tensile_CREATE_COMMAND: /home/lamikr/own/fedora/rocm/src/sdk/rocm_sdk_builder_612/builddir/023_02_rocBLAS/virtualenv/lib/python3.11/site-packages/Tensile/bin/TensileCreateLibrary;--merge-files;--separate-architectures;--lazy-library-loading;--no-short-file-names;--no-library-print-debug;--code-object-version=V5;--cxx-compiler=hipcc;--jobs=10;--library-format=msgpack;--architecture=gfx1102;/home/lamikr/own/fedora/rocm/src/sdk/rocm_sdk_builder_612/src_projects/rocBLAS/library/src/blas3/Tensile/Logic/asm_full;/home/lamikr/own/fedora/rocm/src/sdk/rocm_sdk_builder_612/builddir/023_02_rocBLAS/Tensile;HIP
-- Tensile_MANIFEST_FILE_PATH: /home/lamikr/own/fedora/rocm/src/sdk/rocm_sdk_builder_612/builddir/023_02_rocBLAS/Tensile/library/TensileManifest.txt
'/home/lamikr/own/fedora/rocm/src/sdk/rocm_sdk_builder_612/builddir/023_02_rocBLAS/virtualenv/lib/python3.11/site-packages/Tensile/bin/TensileCreateLibrary' '--merge-files' '--separate-architectures' '--lazy-library-loading' '--no-short-file-names' '--no-library-print-debug' '--code-object-version=V5' '--cxx-compiler=hipcc' '--jobs=10' '--library-format=msgpack' '--architecture=gfx1102' '/home/lamikr/own/fedora/rocm/src/sdk/rocm_sdk_builder_612/src_projects/rocBLAS/library/src/blas3/Tensile/Logic/asm_full' '/home/lamikr/own/fedora/rocm/src/sdk/rocm_sdk_builder_612/builddir/023_02_rocBLAS/Tensile' 'HIP' '--generate-manifest-and-exit'
Traceback (most recent call last):
  File "/home/lamikr/own/fedora/rocm/src/sdk/rocm_sdk_builder_612/builddir/023_02_rocBLAS/virtualenv/lib/python3.11/site-packages/Tensile/bin/TensileCreateLibrary", line 30, in <module>
    from Tensile.TensileCreateLibrary import TensileCreateLibrary
ModuleNotFoundError: No module named 'Tensile.TensileCreateLibrary'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/lamikr/own/fedora/rocm/src/sdk/rocm_sdk_builder_612/builddir/023_02_rocBLAS/virtualenv/lib/python3.11/site-packages/Tensile/bin/TensileCreateLibrary", line 37, in <module>
    from Tensile.TensileCreateLibrary import TensileCreateLibrary
ModuleNotFoundError: No module named 'Tensile.TensileCreateLibrary'
CMake Error at /home/lamikr/own/fedora/rocm/src/sdk/rocm_sdk_builder_612/builddir/023_02_rocBLAS/virtualenv/cmake/TensileConfig.cmake:277 (message):
  Error creating Tensile library: 1
Call Stack (most recent call first):
  library/src/CMakeLists.txt:74 (TensileCreateLibraryFiles)

```

**Comment by daniandtheweb** [comment]
I think that should be better added as an issue as that would be easier to track. The same should go for the triton cache issue as it's not immediate to find the solution in the closed issue.

**Comment by lamikr** [comment]
Yeah, the documentation needs improving. Some kind of faq/troupleshooting section that is separate from the main README

**Comment by lamikr** [comment]
Python 3.11 should be now good to go, thank you @daniandtheweb  for creating the patches!

**Pull request merged**